### PR TITLE
Extract common part of relay loops

### LINKED ISSUE #161: Relay can't detect if node is offline
Part1: It looks like after we have migrated to shared client, we also lost ability to receive transport-related RPC errors. I.e. if you start relay with both nodes online, but then stop and start some node again, relay will hang forever, because transport errors are lost somewhere in background threads.

Part2 (depends on way how we resolve part1): if stopping node would spoil `jsonrpsee::Client` forever, we'll need to create new instance. I've started implementing that before realizing that all transport errors are lost. Intermediate results are [in this branch](https://github.com/paritytech/parity-bridges-common/tree/reconnect-on-transport-error). I'll leave it there for now.

### PR DESCRIPTION
**Pull request by svyatonik**
closes #161
required for #605

While this change seems massive, it is not actually. There are 3 main changes:
1) new `relays/utils/src/relay_loop.rs` file that has single `run` function. This functions is responsible for reconnecting to failed clients and calling actual loop function;
2) new basic `Client` trait in that only has `fn reconnect()` and `type Error;` members;
3) rebasing all existing loops to this new function and trait. This has caused most "visible changes", though they're mostly changes to tab level.

Later this basic `run` function will be extended to e.g. spin up metrics and etc.

To answer possible question: why simply not to hide this error processing in client function(s). I.e. instead of propagating this function to relay loop, simply keep retrying where error has happened? 2/3 of relay loops are async and have their own timers, timeouts, ... So if one of clients will keep retrying in one of its functions, loop may still keep e.g. submitting obsolete data to other node. Additionally, some client functions are used outside of loops, where retrying isn't necessary.

### CODE REVIEW

**Review by HCastano — APPROVED** [review]
Looks okay, although I'd wait for another review before merging.

For the future, I would've also appreciated the "rebasing" changes in a different commit(s) from the main changes - it made it a little hard to review.

File: relays/ethereum-client/src/client.rs
Comment (svyatonik) [reply relays/ethereum-client/src/client.rs:56]: Yes - at least in our version of `jsonrpsee`. Maybe in v2 something has changed, but in v1 background thread was just stopped in case of error => client should have been restarted

File: relays/ethereum/src/ethereum_exchange.rs
Comment (svyatonik) [reply relays/ethereum/src/ethereum_exchange.rs:132]: `Error` has moved to base trait definition, so you should have something like `<Self as RelayClient>::Error` now. That's why all changes like that in this PR. I can create a type alias, but not sure if it worth it.

File: relays/exchange-relay/src/exchange_loop.rs
Comment (svyatonik) [reply relays/exchange-relay/src/exchange_loop.rs:30]: Actually backoff was still in headers+messages loops. Here it was removed because exchage loop is not like the other loops - it isn't fully async. It is async from Rust pov, but it has not simultaneous requests to source/target nodes and any internal buffers, so it's more like linear. But I've returned backoff for non-connection errors. I agree that removal was unjustified here.

Please note that I'm not using backoff mechanism for connection-related errors in all loops intentionally. That's because connection-related errors are not creating (at least it was assumed in the design) additional load on client node. So imagine your network is dropped for 10m. In this case backoff mechanism may take another 5m to retry connection. This all will lead in total >15m delay of all bridge-related applications. Delay 10m10s is obviously better than that, given that it doesn't create additional load on target node.

File: relays/ethereum/src/ethereum_exchange.rs
Comment (HCastano) [reply relays/ethereum/src/ethereum_exchange.rs:132]: It's fine, don't worry about it

**Review by tomusdrw — APPROVED** [review]
lgtm! Clarfications on the `Clone` requirement would be good.

**Comment by HCastano** [comment]
@svyatonik pinging you in case you missed Tomek's comments

File: relays/ethereum/src/ethereum_exchange.rs
Comment (svyatonik) [reply relays/ethereum/src/ethereum_exchange.rs:123]: It comes from [here](https://github.com/paritytech/parity-bridges-common/pull/660/files#diff-d8ddb5e795d8db0a4fdf966079e4f1b7e3939391b6d6f44242ebd1b3bc91e036R53). You need to pass client to closure that returns future (i.e. async function) + at the same time leave it in the 'main' function for possible future reconnect.

File: relays/utils/src/relay_loop.rs
Comment (svyatonik) [reply relays/utils/src/relay_loop.rs]: Changed to `&mut self`

**Review by tomusdrw — APPROVED** [review]

**Pull request merged**
